app/src/main/java/org/simple/clinic/patient/PatientRepository.kt

```diff
@@ -175,43 +146,71 @@ class PatientRepository @Inject constructor(
         }
   }
 
-  fun updatePatientStatusToDead(patientUuid: UUID): Completable{
+  fun updatePatientStatusToDead(patientUuid: UUID): Completable {
     return Completable.fromAction {
       database.patientDao()
           .updatePatientStatus(patientUuid, PatientStatus.DEAD)
-      updatePatientsSyncStatus(listOf(patientUuid), PENDING)
+      setSyncStatus(listOf(patientUuid), PENDING)
     }
   }
 
-  fun mergeWithLocalData(serverPayloads: List<PatientPayload>): Completable {
-    return serverPayloads
-        .toObservable()
-        .filter { payload ->
-          val localCopy = database.patientDao().getOne(payload.uuid)
-          localCopy?.syncStatus.canBeOverriddenByServerCopy()
+  override fun recordsWithSyncStatus(syncStatus: SyncStatus): Single<List<PatientSaveModel>> {
+    return database.patientDao()
+        .recordsWithSyncStatus(syncStatus)
+        .firstOrError()
+        .map { results ->
+          results.asSequence()
+              .groupBy { it.patient.uuid }
+              .map { (_, patientQueryModels) ->
+                PatientSaveModel(
+                    patient = patientQueryModels.first().patient,
+                    address = patientQueryModels.first().address,
+                    phoneNumbers = patientQueryModels
+                        .takeIf { it.first().phoneNumber != null }
+                        ?.asSequence()?.map { it.phoneNumber!! }?.toList()
+                        ?: emptyList()
+                )
```

Comment:
As you said this is indeed really complicated. There's a lot happening here. I'd like us to simplify this. I'll try doing this tomorrow.

Reply:
Could you take another look at it? I have taken a stab at simplifying it.

Reply:
Cleaned up very well by making this entire implementation internal to the Patient Dao.